// @mswjs/interceptors/lib/node/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunk42632LKHjs = require('./chunk-42632LKH.js');



var _chunkLK6DILFKjs = require('./chunk-LK6DILFK.js');


var _chunkIDEEMJ3Fjs = require('./chunk-IDEEMJ3F.js');








var _chunkYGM3BCJUjs = require('./chunk-YGM3BCJU.js');

// src/utils/getCleanUrl.ts
function getCleanUrl(url, isAbsolute = true) {
  return [isAbsolute && url.origin, url.pathname].filter(Boolean).join("");
}













exports.BatchInterceptor = _chunk42632LKHjs.BatchInterceptor; exports.INTERNAL_REQUEST_ID_HEADER_NAME = _chunkYGM3BCJUjs.INTERNAL_REQUEST_ID_HEADER_NAME; exports.IS_PATCHED_MODULE = _chunkIDEEMJ3Fjs.IS_PATCHED_MODULE; exports.Interceptor = _chunkYGM3BCJUjs.Interceptor; exports.InterceptorReadyState = _chunkYGM3BCJUjs.InterceptorReadyState; exports.createRequestId = _chunkYGM3BCJUjs.createRequestId; exports.decodeBuffer = _chunkLK6DILFKjs.decodeBuffer; exports.deleteGlobalSymbol = _chunkYGM3BCJUjs.deleteGlobalSymbol; exports.encodeBuffer = _chunkLK6DILFKjs.encodeBuffer; exports.getCleanUrl = getCleanUrl; exports.getGlobalSymbol = _chunkYGM3BCJUjs.getGlobalSymbol; exports.isResponseWithoutBody = _chunkYGM3BCJUjs.isResponseWithoutBody;
